File: packages/agent/agent/agent.py
import structlog
from typing import Awaitable, Callable
from .provider.openai_types import (
    ToolRequestMessage,
    ToolFunction,
    ToolDefinition,
    Message,
)
from .tool_manager import ToolManager

log = structlog.get_logger(__name__)
MakeLLMRequest = Callable[[list[Message], list[ToolDefinition]], Awaitable[Message]]


TASK_COMPLETE_TOOL = ToolDefinition(
    type="function",
    function=ToolFunction(
        name="task_complete",
        description="Call when there is nothing more to do on this task, either because it's done or because because the task is not possible.",
        parameters={
            "type": "object",
            "properties": {},
        },
    ),
)


class Agent:
    def __init__(self, make_llm_request: MakeLLMRequest, tool_manager: ToolManager):
        self.make_llm_request = make_llm_request
        self.tool_manager = tool_manager

        self.messages: list[Message] = []

    async def run(
        self, initial_messages: list[Message], max_iter: int = 100
    ):
        self.messages = initial_messages

        for _ in range(max_iter):
            tools = self.tool_manager.get_tools() + [TASK_COMPLETE_TOOL]
            log.info("querying_llm", tools=[t.function.name for t in tools])
            message = await self.make_llm_request(
                self.messages,
                tools,
            )
            log.info("llm_response_received")

            self.messages.append(message)

            stop = None
            if isinstance(message, ToolRequestMessage):
                # Do them sequentially so that the LLM get confused by concurrency issues
                results: list[Message] = []
                for call in message.tool_calls:
                    if call.function.name == TASK_COMPLETE_TOOL.function.name:
                        stop = 'stop_tool'
                        break
                    else:
                        result = await self.tool_manager.call_tool(call)
                        results.append(result)

                self.messages.extend(results)
            else:
                if message.content == "":
                    # No content, we are probably done
                    stop = 'empty_content'
        
            if stop:
                log.info("task_complete", reason=stop)
                break

File: simple-agent/src/agent.rs
use std::{collections::HashMap, sync::Arc};
use log::{info, warn};
use crate::llm_api::types::{Message, ToolCall, ToolDefinition, ToolFunction, ToolFunctionType};
use super::llm_client::LLMClient;
use super::tool_manager::ToolManager;

pub struct Agent {
    llm_client: LLMClient,
    tool_manager: ToolManager,
    pub messages: Vec<Message>,

    task_complete_tool: ToolDefinition
}


impl Agent {
    pub fn new(llm_client: LLMClient, tool_manager: ToolManager) -> Self {
        Self {
            llm_client,
            tool_manager,
            messages: Vec::new(),

            task_complete_tool: ToolDefinition {
                type_: ToolFunctionType::Function,
                function: ToolFunction {
                    name: "task_complete".to_string(),
                    description: Some("Call when there is nothing more to do on this task, either because it's done or because the task is not possible.".to_string()),
                    parameters: HashMap::from_iter([
                        (
                            "type".to_string(),
                            serde_json::Value::String("object".to_string()),
                        ),
                        (
                            "properties".to_string(),
                            serde_json::Value::Object(serde_json::map::Map::new()),
                        ),
                    ]),
                }
            },
        }
    }

    pub async fn run(&mut self, initial_messages: Vec<Message>, max_iter: usize) -> Result<(), Box<dyn std::error::Error + Send + Sync>> {
        self.messages.clear();
        self.messages.extend(initial_messages);

        for _ in 0..max_iter {
            let mut tools = self.tool_manager.get_tools().await;
            tools.push(self.task_complete_tool.clone());

            info!("querying_llm");

            let chat_response = self.llm_client.make_request(self.messages.clone(), Some(tools)).await?;
            let message = &chat_response.choices[0].message;

            info!("llm_response_received");
            self.messages.push(message.clone());

            let mut stop = None;

            match message {
                Message::TextMessage(text_message) => {
                    if text_message.content.is_empty() {
                        stop = Some("empty_content".to_string());
                    }
                }
                Message::ToolRequestMessage(tool_request_message) => {
                    let mut results = Vec::new();
                    for call in &tool_request_message.tool_calls {
                        if call.function.name == self.task_complete_tool.function.name {
                            stop = Some("stop_tool".to_string());
                            break;
                        } else {
                            let result = self.tool_manager.call_tool(call.clone()).await;
                            results.push(result);
                        }
                    }
                    self.messages.extend(results);
                }
                Message::ToolResponseMessage(tool_response_message) => {
                    warn!("got_tool_response_from_llm")
                }

            }

            if let Some(reason) = stop {
                info!("task_complete");
                return Ok(())
            }
        }

        return Ok(());
    }
}
